File: ios/ClipboardListenerModule.swift
import ExpoModulesCore

public class ClipboardListenerModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ClipboardListener')` in JavaScript.
    Name("ClipboardListener")

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    //Registers a notification observer for clipboard changes.
    OnStartObserving {
      NotificationCenter.default.addObserver(
        forName: UIPasteboard.changedNotification,
        object: nil,
        queue: nil
      ) { _ in
        self.sendEvent("onChange", ["value": UIPasteboard.general.string ?? ""])
      }
    }

    // Removes the observer when no longer needed.
    OnStopObserving {
      NotificationCenter.default.removeObserver(
        self, name: UIPasteboard.changedNotification, object: nil)
    }

    // Sets the clipboard value.
    Function("setString") { (value: String?) in
      if value != nil {
        UIPasteboard.general.string = value
      }
    }


    Function("getString") { () -> String in
      return UIPasteboard.general.string ?? ""
    }

  }
}

File: android/src/main/java/expo/modules/clipboardlistener/ClipboardListenerModule.kt
package expo.modules.clipboardlistener

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL
import android.content.ClipboardManager
import android.content.Context
import android.content.ClipData

class ClipboardListenerModule : Module() {
  private lateinit var clipboardManager: ClipboardManager
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ClipboardListener')` in JavaScript.
    Name("ClipboardListener")

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    // Initializes the ClipboardManager and adds a listener for clipboard changes.
    OnCreate {
      val context = appContext.reactContext
      clipboardManager = context?.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
      clipboardManager.addPrimaryClipChangedListener {
        sendEvent("onChange", mapOf<String, String>(
          "value" to (clipboardManager.primaryClip?.getItemAt(0)?.text?.toString() ?: "")
        ))
      }
    }

    // Cleans up the listener to prevent memory leaks.
    OnDestroy {
      clipboardManager.removePrimaryClipChangedListener { /* No-op listener for removal */ }
    }

    // Sets the clipboard value.
    Function("setString") { value: String? ->
      if (value == null) {
        return@Function
      }
      val clip = ClipData.newPlainText(null, value)
      clipboardManager.setPrimaryClip(clip)
    }

    // Gets the clipboard value.
    Function("getString") { 
      clipboardManager.primaryClip?.getItemAt(0)?.text?.toString() ?: ""
    }
  }
}